### PR TITLE
Rewire INSTR before EXPAND-TO-NATIVE-INSTRUCTIONS in COST-FUNCTION

### LINKED ISSUE #458: cnot-to-native-cnots does not guard against a path length between q1 and q0 being zero
And it may thus enter an infiniloop.

### LINKED ISSUE #511: COST-FUNCTION on non-native instruction can cause the compiler to loop
On a chip with live qubits 3 and 4, witness `CZ 0 1`.  This bug was introduced by the fidelity-addresser merge.

### PR DESCRIPTION
**Pull request by notmgsk**
`cost-function` expands its input instruction into a list of native instructions. If the instruction is not first rewired, then `expand-to-native-instructions` will get confused and infiniloop. For example, if a chip has live physical qubits 3 and 4, then the instruction `CZ 0 1` will be expanded into `CAN ... 0 1` will be expanded into `CZ 0 1` will be expanded into `CAN ... 0 1` ... etc.

My first thought was that `expand-to-native-instructions` should be aware of the current rewiring by providing it a `addresser-state` object containing the rewiring, or even just the rewiring itself. Until now, however, that hasn't been required and so maybe isn't the best solution.

Second (and current) solution, is to temporarily apply the current rewiring to the instruction in `cost-function` (being careful not to mutate the instruction beyond the invocation of `expand-to-...`).

Closes #511 and closes #458.

### CODE REVIEW

**Comment by notmgsk** [comment]
Latest commit tries to apply a rewiring if one exists.

**Comment by notmgsk** [comment]
I've tried to convince myself that I'm not breaking anything, and having learned a good deal about the addresser along the way: I'm not 100% but relatively confident.

File: tests/compilation-tests.lisp [hunk 1]
```diff
@@ -167,3 +167,7 @@ CNOT 0 2"))
     ;; NOTE: Decomposing into fewer 2q gates is more of a regression
     ;; test on quality of compilation, and not on correctness.
     (is (>= 6 (length 2q-code)))))
+
+(deftest test-sohaib-fidelity-rewiring-regression ()
+  (is (compiler-hook (parse "CZ 0 1")
```
Comment (stylewarning) [inline tests/compilation-tests.lisp]: do you want `is` or `not-signals`?

File: src/addresser/rewiring.lisp [hunk 2]
```diff
@@ -309,3 +309,13 @@ BODY as an implicit PROGN."
               first-rewiring-string second-rewiring-string)
       (values (make-rewiring-from-string first-rewiring-string)
               (make-rewiring-from-string second-rewiring-string)))))
+
+(defun rewiring-available-p (rewiring lq)
```
Comment (stylewarning) [inline src/addresser/rewiring.lisp]: I prefer `rewiring-assigned-for-qubit-p`

File: src/addresser/rewiring.lisp [hunk 3]
```diff
@@ -309,3 +309,13 @@ BODY as an implicit PROGN."
...
+  "Test whether REWIRING contains a non-nil rewiring for logical qubit LQ."
+  (aref (rewiring-l2p rewiring) lq))
+
+(defun rewiring-available-for-instruction-p (rewiring instr)
```
Comment (stylewarning) [inline src/addresser/rewiring.lisp]: similarly, `rewiring-assigned-for-instruction-qubits-p`

File: src/addresser/rewiring.lisp [hunk 4]
```diff
@@ -309,3 +309,13 @@ BODY as an implicit PROGN."
...
+  "Test whether every logical qubit in INSTR has been rewired in REWIRING."
+  (every (a:curry #'rewiring-available-p rewiring)
```
Comment (stylewarning) [inline src/addresser/rewiring.lisp]: why not go whole hog and

```
(every (a:compose (a:curry ...) #'qubit-index)
       (application-arguments instr))
```

File: tests/compilation-tests.lisp (hunk 1 above)
Comment (notmgsk) [reply tests/compilation-tests.lisp]: I'm not sure actually. The failure mode is either an infinite loop which blows the stack or memory exhaustion.

Comment (stylewarning) [reply tests/compilation-tests.lisp]: i would add a timeout with bordeaux-threads and a `not-signals` then

File: src/addresser/rewiring.lisp (hunk 4 above)
Comment (notmgsk) [reply src/addresser/rewiring.lisp]: now you're cooking with functions

File: src/addresser/rewiring.lisp [hunk 5]
```diff
@@ -309,3 +309,13 @@ BODY as an implicit PROGN."
...
       (values (make-rewiring-from-string first-rewiring-string)
               (make-rewiring-from-string second-rewiring-string)))))
+
+(defun rewiring-assigned-for-qubit-p (rewiring lq)
+  "Test whether REWIRING contains a non-nil rewiring for logical qubit LQ."
+  (aref (rewiring-l2p rewiring) lq))
+
+(defun rewiring-assigned-for-instruction-qubits-p (rewiring instr)
+  "Test whether every logical qubit in INSTR has been rewired in REWIRING."
+  (every (a:compose #'qubit-index
```
Comment (stylewarning) [inline src/addresser/rewiring.lisp]: is this what you want?

```
CL-USER> (funcall (alexandria:compose #'1+ #'sqrt) 3)
2.732051
CL-USER> (funcall (alexandria:compose #'sqrt #'1+) 3)
2.0
```

File: tests/compilation-tests.lisp (hunk 1 above)
Comment (stylewarning) [reply tests/compilation-tests.lisp]: still this

Comment (notmgsk) [reply tests/compilation-tests.lisp]: Give me chance!

File: tests/compilation-tests.lisp [hunk 6]
```diff
@@ -169,5 +169,7 @@ CNOT 0 2"))
     (is (>= 6 (length 2q-code)))))
 
 (deftest test-sohaib-fidelity-rewiring-regression ()
-  (is (compiler-hook (parse "CZ 0 1")
-                     (quil::build-chip-from-digraph '((2 3) (3 2))))))
+  (not-signals bt:timeout
+    (bt:with-timeout (1)
+      (is (compiler-hook (parse "CZ 0 1")
```
Comment (stylewarning) [inline tests/compilation-tests.lisp]: nix `is`

**Pull request merged**